// react-promise-tracker/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "trackPromise", {
  enumerable: true,
  get: function get() {
    return _trackPromise.trackPromise;
  }
});
Object.defineProperty(exports, "manuallyResetPromiseCounter", {
  enumerable: true,
  get: function get() {
    return _trackPromise.manuallyResetPromiseCounter;
  }
});
Object.defineProperty(exports, "manuallyDecrementPromiseCounter", {
  enumerable: true,
  get: function get() {
    return _trackPromise.manuallyDecrementPromiseCounter;
  }
});
Object.defineProperty(exports, "manuallyIncrementPromiseCounter", {
  enumerable: true,
  get: function get() {
    return _trackPromise.manuallyIncrementPromiseCounter;
  }
});
Object.defineProperty(exports, "promiseTrackerHoc", {
  enumerable: true,
  get: function get() {
    return _trackerHoc.promiseTrackerHoc;
  }
});
Object.defineProperty(exports, "usePromiseTracker", {
  enumerable: true,
  get: function get() {
    return _trackerHook.usePromiseTracker;
  }
});

var _trackPromise = require("./trackPromise");

var _trackerHoc = require("./trackerHoc");

var _trackerHook = require("./trackerHook");
